src/main/scala/analysis/SummaryGenerator.scala

```diff
@@ -0,0 +1,186 @@
+      //case variable: LocalStackVariable => None
+      case variable: UnknownMemory => Some(FalseBLiteral)
+    }
+  }
+
+  private def toExprGamma(variable: Taintable): Option[BExpr] = {
+    variable match {
+      case variable: GlobalVariable => Some(BinaryBExpr(BoolOR, variable.toGamma, variable.L))
+      case _ => toGamma(variable)
+    }
+  }
+
+  /**
+   * Get a map of variables to variables which have tainted it in the procedure.
+   */
+  private def getTainters(procedure: Procedure, variables: Set[Taintable]): Map[Taintable, Set[Taintable]] = {
+    varDepsSummaries.getOrElse(procedure, Map())
+  }
+
+  /**
+   * Generate requires clauses for a procedure. Currently this does nothing.
+   */
+  def generateRequires(procedure: Procedure): List[BExpr] = List()
+
+  /**
+   * Generate ensures clauses for a procedure. Currently, all generated ensures clauses are of the form (for example)
+   *   ensures Gamma_R2 || (Gamma_R2 == old(Gamma_y)) || (Gamma_R2 == old(Gamma_R1)) || (Gamma_R2 == old(Gamma_R2));
+   * whenever this can be done soundly.
+   */
+  def generateEnsures(procedure: Procedure): List[BExpr] = {
+    if procedure.blocks.isEmpty then return List()
+
+    // We only need to make ensures clauses about the gammas of modified variables.
+    val relevantVars = variables.filter { v =>
+      v match {
+        case v: Global => procedure.modifies.contains(v)
+        case _ => true
+      }
+    }
+
+    // Use rnaResults to find stack function arguments
+    val tainters = relevantVars.map {
+      v => (v, Set())
+    }.toMap ++ getTainters(procedure, variables ++ rnaResults(procedure.begin) + UnknownMemory()).filter { (variable, taints) =>
+      relevantVars.contains(variable)
+    }
+
+    tainters.toList
+      .flatMap { (variable, taints) =>
+      {
+        // If our variable was tainted by memory that we know nothing about, it is sound to assume that we
+        // know nothing about its gamma in the post state.
+        if taints.contains(UnknownMemory()) then None
+        else toGamma(variable).flatMap
+          { varGamma => {
+            if taints.isEmpty then {
+              // If the variable is tainted by nothing, it must have been set to some constant value, meaning
+              // its gamma is low. Note that if a variable was unchanged, it would be tainted by itself.
+              Some(varGamma)
+            } else {
+              // Else, we must consider the variables which have tainted this variable. We can soundly assert
+              // that this variable's gamma must be equal to one of its tainter's gammas, or it is set to
+              // low.
+              //
+              // In a correct procedure, we never have L(x) = true and Gamma_x = false, so we can safely
+              // assume L(x) || Gamma_x == Gamma_x. This means that the gamma of any expression is equal
+              // to the conjunction of the gammas of the variables and the loads in the expression.
```

Comment:
This doesn't seem necessarily correct to me. Since a variable's security level can change throughout a procedure, its final Gamma value isn't necessarily going to directly relate to the input Gamma values of one of the variables that tainted it?

For example, if we have R0 and R1 as low and R2 as high at the start of a procedure and then the body of the procedure is the following:
```
R0 := R2 + 1
R1 := R0 + 1
```
Would the analysis be able to tell that R1's gamma value actually depends on the initial gamma value of R2, not R1?

Reply:
> Since a variable's security level can change throughout a procedure, its final Gamma value isn't necessarily going to directly relate to the input Gamma values of one of the variables that tainted it?

Maybe I've missed something, but I thought that a change in security level doesn't change the gamma of any variables

> Would the analysis be able to tell that R1's gamma value actually depends on the initial gamma value of R2, not R0?

It will be able to tell, yes. After the first assignment, R0 will depend on {R2}, and after the second assignment, R1 will depend on whatever R0 depends on, so {R2}.

Reply:
I was using 'security level' and 'gamma' interchangeably there, not referring to the security policy for a variable, sorry for the confusion.